src/sql_engine/src/query/mod.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright 2020 Alex Dukhno
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+///! Module for representing how a query will be executed and values represented
+///! during runtime.
+
+mod plan;
+mod repr;
+mod relation;
+mod scalar;
+mod transform;
+
+pub use scalar::{ScalarOp};
+pub use repr::{Datum, Row};
+pub use plan::Plan;
+pub use transform::QueryTransform;
+
+use sql_types::SqlType;
+
+/// A type of a column
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ColumnType {
+    #[allow(dead_code)]
+    nullable: bool,
+    /// the sql type
+    sql_type: SqlType
+}
+
+impl ColumnType {
+    pub fn new(sql_type: SqlType) -> Self {
+        Self {
+            nullable: false,
+            sql_type
+        }
+    }
+
+    pub fn typ(&self) -> SqlType {
+        self.sql_type
+    }
+}
+
+/// relation (table) type
+/// A relation type is just the types of the columns.
+/// Materialize uses this same concept.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct RelationType {
+    /// the types of the columns in the specified order.
+    column_types: Vec<ColumnType>,
+    // Materialized also has a Vec<Vec<usize>> to represent the indices
+    // available for this table but I do not know how that is going to work
+    // in this database so I am leaving it out.
+}
+
+impl RelationType {
+    pub fn new(column_types: Vec<ColumnType>) -> Self {
+        Self { column_types }
+    }
+
+    pub fn columns(&self) -> &[ColumnType] {
+        self.column_types.as_slice()
+    }
+}
+
+// this works for now, but ideally this should be usize's instead of strings.
+
+/// represents a table uniquly
+///
+/// I would like this to be a single 64 bit number where the top bits are the
+/// schema id and lower bits are the table id.
+#[derive(Debug, Clone, Hash, Eq, PartialEq, Ord, PartialOrd)]
+pub struct TableId(SchemaId, String);
```

Comment:
I have checked on rust [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=153a52fcd2da0086d7804aef0e8aaf63) the memory size of `u64` and `(u32, u32)` is the same, so I'd go with tuple struct instead of byte manipulation logic
and we can give them type aliases like `type TableId = u32`

Reply:
I do not think we are in a state to use this yet. We need a way to lookup a schema or table name to retrieve its id. I also thought maybe the storage could be based on these ids instead of the actual names, but I am not sure how the actual names would be stored when writing to a file.